feat: implemented with_field() for FixedSizeListBuilder

istvan-fodor opened the pull request:
# Which issue does this PR close?

Closes #1353 

# Rationale for this change
 
`FixedSizeListBuilder` doesn't allow users to specify the inner field of elements and defaults to the following FieldRef: `Arc::new(Field::new("item", values_data.data_type().clone(), true))`

We would like to keep the default behavior as-is, but also allow users to specify use their own FieldRef. 

# What changes are included in this PR?

1. Added `with_field` to `FixedSizeListBuilder`, copied over from `GenericListBuilder`
2. Added new `field: Option<FieldRef>` to `FixedSizeListBuilder`
3. Implemented field logic patterned after `GenericListBuilder` in `finish()` and `finish_build()`
4. Changed unsafe build to safe build (see user-facing changes)
5. Unit tests for 3 scenarios:
    1. `with_field` used
    2. `with_field` used with non-nullable field, expect panic
    3. `with_field` used with wrong type, expect panic  

# Are there any user-facing changes?

Yes.

The `finish_cloned()` and `finish()` methods no longer use the `build_unchecked()` method, but rely on `build()`. If the build fails, the thread panics.

alamb reviewed (commented):
Thank you @istvan-fodor  -- this looks like a nice API improvement to me.

alamb commented on arrow-array/src/builder/fixed_size_list_builder.rs:
```diff
@@ -194,17 +212,23 @@ where
-        .add_child_data(values_data)
-        .nulls(nulls);
 
-        let array_data = unsafe { array_data.build_unchecked() };
```
I don't understand the rationale for this change.  Is there some case where invalid arrays are created?

Switching to use build() means the data is now checked twice (once when added to the builder and once when the array is built)

If there are ways to construct an invalid FIxedListArray via the builder APIs, I think those should be fixed  rather than running the expensive `build()` checks again.

For example, if you are trying to validate that a builder with a field that was declared to be non-null actually has no nulls, it would be much faster to leave the `build_unchecked()` call and do a separate check that `array_data.null_count()` was zero

istvan-fodor replied:
The rationale is that with giving the ability to the user to change the field we use, the values_builder and the field can be out of sync.  

If we use build_unchecked, the build will not fail as expected, so the unit tests 
    `builder::fixed_size_list_builder::tests::test_fixed_size_list_array_builder_with_field_null_panic` and
    `builder::fixed_size_list_builder::tests::test_fixed_size_list_array_builder_with_field_type_panic`will fail. 
    
 I think the best place would be to do this check in `with_field`, asserting that the nullable and data_type are the same as the values-builder, but I didn't find a way to actually get the values from values_builder. 
  
  If we stick with `build_unchecked`, we could have this check once the `values_builder` is built (data_type and nullable are readily available on the built `ArrayData`).
  
  I am updating the PR with this version and let me know if this is better.

alamb replied:
Looks good 👍

alamb commented on arrow-array/src/builder/fixed_size_list_builder.rs:
```diff
@@ -247,6 +298,180 @@ mod tests {
+            true,
+        ));
+
+        //  [[0, 1, 2], null, [3, null, 5], [6, 7, null]]
+        builder.values().append_value(0);
+        builder.values().append_value(1);
+        builder.values().append_value(2);
+        builder.append(true);
+        builder.values().append_null();
+        builder.values().append_null();
+        builder.values().append_null();
+        builder.append(false);
+        builder.values().append_value(3);
+        builder.values().append_null();
+        builder.values().append_value(5);
+        builder.append(true);
+        builder.values().append_value(6);
+        builder.values().append_value(7);
+        builder.values().append_null();
+        builder.append(true);
```
It might be nice to avoid repeating this same setup multiple times in the tests (and it would make it clear the tests share the same data)

for example
```rust
fn make_list_builder() -> FixedSizedListBuilder<i32> { 
  ...
}
```


```suggestion
        let builder = make_list_builder();
```

istvan-fodor replied:
Makes sense, I'll go through the tests and rework the builders.

alamb reviewed (approved):
Thank you @istvan-fodor -- I think this looks good to me. I had a small test organization suggestion but I don't think it is required to merge this PR

I don't think the CI failures are related to changes in this PR -- I filed https://github.com/apache/arrow-rs/pull/5544 to fix. 

Thanks again

cc @TimDiekmann

istvan-fodor commented:
@alamb , not sure you saw my comment on line 328. Can you check? The new behavior of non-nullable fields raises an additional question about null fields and I wanted to make sure this is in line with the rest of the arrow builders.

tustvold commented:
I could be missing something but I don't see a comment on line 328? Is it possible that it is pending and the review hasn't been submitted?

istvan-fodor commented on arrow-array/src/builder/fixed_size_list_builder.rs:
```diff
@@ -248,9 +325,37 @@ mod tests {
     #[test]
-    fn test_fixed_size_list_array_builder_finish_cloned() {
+    fn test_fixed_size_list_array_builder_with_field() {
+        let builder = make_list_builder(false, false);
```
@alamb , I had a question about this test in particular. With the new capability to set the field as a non-nullable we have the situation where the builder has a null element built like this: 

```rust
 builder.values().append_null();
 builder.values().append_null();
 builder.values().append_null();
 builder.append(false);
```

If the field is set to non-nullable, this element will fail the null check. I wanted to check with you if this is logically right or not? Are we making a distinction between nulls in the values builder, where let's say 1 of the 3 elements of a particular fixed size list is null (which would be a valid fail on non-nullability) or nulls in the main builder (which is built up as 3 nulls + a `append(false)` in the values builder). To me the second one is questionable and comes down to how you want this to work. If this test should not fail with the second complete null case, then I have to go back and revise the null check of the values builder and somehow control for null elements of the main builder.

tustvold replied:
The logic we apply for FixedSizeList and StructArray elsewhere is to permit "masked" nulls, this is because a null must always take up space. I am not sure if we want to replicate this logic here

alamb replied:
> If the field is set to non-nullable, this element will fail the null check. I wanted to check with you if this is logically right or not?

I think I would expect it to be consistent with whatever `FixedSlizeListArray::try_new` did

istvan-fodor commented:
> I could be missing something but I don't see a comment on line 328? Is it possible that it is pending and the review hasn't been submitted?

Yes, that was the case. I just changed it to a regular comment

SteampunkIslande commented:
Hello ! I'm very interested in this PR, I really need FixedSizeList columns and didn't find a workaround that makes the FixedSizeListBuilder create a column with an appropriate name. So I'm really looking forward for this to be accepted

tustvold commented:
There is currently an ongoing discussion w.r.t null handling, in the short term you could possibly use the array constructors directly - https://docs.rs/arrow-array/latest/arrow_array/array/struct.FixedSizeListArray.html#method.try_new

SteampunkIslande commented:
> There is currently an ongoing discussion w.r.t null handling, in the short term you could possibly use the array constructors directly - https://docs.rs/arrow-array/latest/arrow_array/array/struct.FixedSizeListArray.html#method.try_new

Thanks, but I really need a builder. Tried implementing it myself with no luck so far. I'm pretty surprised that the builder doesn't know its field name though. There should be a way to know right ? At least I'm hoping that with your pull request it will be fixed

alamb commented on arrow-array/src/builder/fixed_size_list_builder.rs:
```diff
@@ -248,9 +325,37 @@ mod tests {
+
+    #[test]
+    #[should_panic(
+        expected = "field is nullable = false, but the values_builder contains null values"
```
this doesn't seem right -- the builder doesn't have null values, instead it has no values at all.

alamb commented on arrow-array/src/builder/fixed_size_list_builder.rs:
```diff
@@ -194,13 +227,31 @@ where
+                );
+                if !f.is_nullable() {
+                    assert!(
+                        values_data.null_count() == 0,
```
I think it is valid to have a zero length array (which would also have no nulls, but the null count would be zero)


```suggestion
                        values_data.null_count() == 0 || values_data.len() == 0,
```

(same for the check below)

alamb reviewed (commented):
Thanks you @istvan-fodor  -- I think this is getting close. Some final touchups:

1. Update the test for zero length lists (should not fail null check)
2. Add a test for building a 2 element array where one elemnt is NULL (with a null element)
3. (maybe) Another test showing the null behavior you are asking about

istvan-fodor commented:
@alamb @tustvold , I added some changes.

1. null check now follows the `FixedSizeListArray::try_new` logic
2. Added two new unit tests to test scenario with nulls: `test_fixed_size_list_array_builder_with_field_and_null` and `test_fixed_size_list_array_builder_cloned_with_field_and_null`
